### PR TITLE
feat(dag): use DAG manifests for faster, dependable registry integration

### PR DESCRIPTION
**Pull request by b5**

### CODE REVIEW

File: actions/dataset.go [hunk 1]
```diff
@@ -242,21 +242,56 @@ func localUpdate(node *p2p.QriNode, ref *repo.DatasetRef, secrets map[string]str
...
+				errs <- pinner.Pin(datastore.NewKey(ref.Path), true)
+			}
+		}()
 	}
 
-	if ref.Path == "" && node != nil {
-		if err := node.RequestDataset(ref); err != nil {
-			return fmt.Errorf("error requesting dataset: %s", err.Error())
+	go func() {
+		errs <- base.FetchDataset(node.Repo, ref, true, true)
```
Comment (dustmop) [inline actions/dataset.go]: Is there a race condition where both tasks could succeed?

File: actions/dataset_ref.go [hunk 2]
```diff
@@ -13,12 +15,45 @@ import (
...
+	if err := repo.CanonicalizeDatasetRef(node.Repo, ref); err == nil && ref.Path != "" {
 		return true, nil
 	} else if err != nil && err != repo.ErrNotFound && err != profile.ErrNotFound {
+		// return early on any non "not found" error
 		return false, err
 	}
 
-	return false, n.ResolveDatasetRef(ref)
+	errs := make(chan error)
+	tasks := 1
```
Comment (dustmop) [inline actions/dataset_ref.go]: Same comment about counting `tasks` as above.

File: actions/dataset_ref.go [hunk 3]
```diff
@@ -13,12 +15,45 @@ import (
...
+		tasks++
+		go func() {
+			if ds, err := rc.GetDataset(ref.Peername, ref.Name, ref.ProfileID.String(), ref.Path); err == nil {
+				ref.Peername = ds.Peername
+				ref.Name = ds.Name
+				ref.ProfileID, _ = profile.IDB58Decode(ds.Commit.Author.ID)
+				ref.Path = ds.Path
+				errs <- nil
+			}
+			errs <- fmt.Errorf("not found")
```
Comment (dustmop) [inline actions/dataset_ref.go]: How about using repo.ErrNotFound?

File: actions/registry.go [hunk 4]
```diff
@@ -81,17 +81,34 @@ func Pin(node *p2p.QriNode, ref repo.DatasetRef) (err error) {
 		}
 	}
 
-	if !node.Online {
-		if err = node.GoOnline(); err != nil {
-			return err
-		}
-	}
+	// if !node.Online {
```
Comment (dustmop) [inline actions/registry.go]: Delete instead of commenting.

File: actions/registry.go [hunk 5]
```diff
@@ -81,17 +81,34 @@ func Pin(node *p2p.QriNode, ref repo.DatasetRef) (err error) {
...
 	var addrs []string
-	for _, maddr := range node.EncapsulatedAddresses() {
-		addrs = append(addrs, maddr.String())
+	// for _, maddr := range node.EncapsulatedAddresses() {
+	// 	addrs = append(addrs, maddr.String())
+	// }
+
+	ng, err := newNodeGetter(node)
+	if err != nil {
+		return nil
```
Comment (dustmop) [inline actions/registry.go]: `return err`?

File: cmd/manifest.go [hunk 6]
```diff
@@ -0,0 +1,173 @@
...
+	"strings"
+
+	"github.com/ghodss/yaml"
+	"github.com/qri-io/dag"
+	"github.com/qri-io/ioes"
+	"github.com/qri-io/qri/lib"
+	"github.com/spf13/cobra"
+)
+
+// NewManifestCommand creates a new `qri search` command that searches for datasets
```
Comment (dustmop) [inline cmd/manifest.go]: Description is inaccurate

File: cmd/manifest.go [hunk 7]
```diff
@@ -0,0 +1,173 @@
...
+func NewManifestCommand(f Factory, ioStreams ioes.IOStreams) *cobra.Command {
+	o := &ManifestOptions{IOStreams: ioStreams}
+	cmd := &cobra.Command{
+		Use:    "manifest",
+		Hidden: true,
+		Short:  "dataset manifest interation",
```
Comment (dustmop) [inline cmd/manifest.go]: "interation" spelling?

File: cmd/manifest.go [hunk 8]
```diff
@@ -0,0 +1,173 @@
...
+	missing.Flags().BoolVar(&o.Pretty, "pretty", false, "print output without indentation, only applies to json format")
+	missing.Flags().BoolVar(&o.Hex, "hex", false, "hex-encode output")
+	missing.Flags().StringVar(&o.File, "file", "", "manifest file")
+
+	cmd.AddCommand(get, missing)
+
+	return cmd
+}
+
+// ManifestOptions encapsulates state for the get command
```
Comment (dustmop) [inline cmd/manifest.go]: for the `manifest` command

File: actions/dataset.go [hunk 9]
```diff
@@ -242,21 +242,56 @@ func localUpdate(node *p2p.QriNode, ref *repo.DatasetRef, secrets map[string]str
...
+			if err != nil {
+				return
+			}
+
+			capi, err := newIPFSCoreAPI(node)
+			if err != nil {
+				return
+			}
+
+			if err := rc.DsyncFetch(node.Context(), ref.Path, ng, capi.Block()); err != nil {
```
[the remark this answers is not in the document]
Comment (ramfox) [reply actions/dataset.go]: Fetching the dataset adds each block to the store as it gets sent over. Happens in the dsync package: https://github.com/qri-io/dag/blob/ef7c42bfdbf94989624bb29dbf3631b37ea2247d/dsync/fetch.go#L124

[the remark this answers is not in the document]
Comment (b5) [reply actions/dataset.go]: yeah, doesn't modify the ref. Instead it interacts directly with the block store that underpins the dataset version specified by ref.

File: actions/dataset.go (hunk 1 above)
Comment (b5) [reply actions/dataset.go]: yes there was. should be fixed now!

**Comment by b5** [comment]
ooooook, think this is ready for another round of review. look ok @ramfox?

**Pull request merged**